x/twap/logic_test.go

```diff
@@ -521,3 +522,421 @@ func (s *TestSuite) TestPruneRecords() {
 
 	s.validateExpectedRecords(expectedKeptRecords)
 }
+
+// TestUpdateRecords tests that the records are updated correctly.
+// It tests the following:
+// - two-asset pools
+// - multi-asset pools
+// - with spot price errors
+// - without spot price errors
+// - that new records are created
+// - older historical records are not updated
+// - spot price error times are either propagated from
+// older records or set to current block time in case error occurred.
+func (s *TestSuite) TestUpdateRecords() {
+	type spOverride struct {
+		poolId      uint64
+		baseDenom   string
+		quoteDenom  string
+		overrideSp  sdk.Dec
+		overrideErr error
+	}
+
+	type expectedResults struct {
+		spotPriceA    sdk.Dec
+		spotPriceB    sdk.Dec
+		lastErrorTime time.Time
+	}
+
+	var spError = errors.New("spot price error")
+
+	validateRecords := func(expectedRecords []expectedResults, actualRecords []types.TwapRecord) {
+		s.Require().Equal(len(expectedRecords), len(actualRecords))
+		for i, r := range expectedRecords {
+			s.Require().Equal(r.spotPriceA, actualRecords[i].P0LastSpotPrice, "record %d", i)
+			s.Require().Equal(r.spotPriceB, actualRecords[i].P1LastSpotPrice, "record %d", i)
+			s.Require().Equal(r.lastErrorTime, actualRecords[i].LastErrorTime, "record %d", i)
+		}
+	}
+
+	tests := map[string]struct {
+		preSetRecords []types.TwapRecord
+		poolId        uint64
+		ammMock       twapmock.ProgrammedAmmInterface
+		spOverrides   []spOverride
+		blockTime     time.Time
+
+		expectedMostRecentRecords []expectedResults
+		expectedHistoricalRecords []expectedResults
+		expectError               error
+	}{
+		"no records pre-set; error": {
+			preSetRecords: []types.TwapRecord{},
+			poolId:        1,
+			blockTime:     baseTime,
+
+			expectError: gammtypes.PoolDoesNotExistError{PoolId: 1},
+		},
+		"existing records in different pool; no-op": {
+			preSetRecords: []types.TwapRecord{baseRecord},
+			poolId:        baseRecord.PoolId + 1,
+			blockTime:     baseTime.Add(time.Second),
+
+			expectError: gammtypes.PoolDoesNotExistError{PoolId: baseRecord.PoolId + 1},
+		},
+		"the returned number of records does not match expected": {
+			preSetRecords: []types.TwapRecord{baseRecord},
+			poolId:        baseRecord.PoolId,
+			blockTime:     baseRecord.Time.Add(time.Second),
+
+			spOverrides: []spOverride{
+				{
+					baseDenom:  baseRecord.Asset0Denom,
+					quoteDenom: baseRecord.Asset1Denom,
+					overrideSp: sdk.NewDec(2),
+				},
+				{
+					baseDenom:  baseRecord.Asset1Denom,
+					quoteDenom: baseRecord.Asset0Denom,
+					overrideSp: sdk.NewDecWithPrec(2, 1),
+				},
+				{
+					baseDenom:  baseRecord.Asset1Denom,
+					quoteDenom: "extra denom B",
```

Comment:
Are spaces allowed in denoms? If not, we should use a valid denom for the extra denom